{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Introduction\n",
    "\n",
    "This notebook demonstrates basic usage of BioThings Explorer, an engine for autonomously querying a distributed knowledge graph. BioThings Explorer can answer two classes of queries -- \"PREDICT\" and \"EXPLAIN\".  PREDICT queries are described in [PREDICT_demo.ipynb](PREDICT_demo.ipynb). Here, we describe EXPLAIN queries and how to use BioThings Explorer to execute them.  A more detailed overview of the BioThings Explorer systems is provided in [these slides](https://docs.google.com/presentation/d/1QWQqqQhPD_pzKryh6Wijm4YQswv8pAjleVORCPyJyDE/edit?usp=sharing).\n",
    "\n",
    "EXPLAIN queries are designed to **identify plausible reasoning chains to explain the relationship between two entities**.  For example, in this notebook, we explore the question:\n",
    "\n",
    "&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;\"*Why does imatinib have an effect on the treatment of chronic myelogenous leukemia (CML)?*\"  \n",
    "\n",
    "Later, we also compare those results to a similar query looking at imatinib's role in treating gastrointestinal stromal tumors (GIST).\n",
    "\n",
    "**To experiment with an executable version of this notebook, [load it in Google Colaboratory](https://colab.research.google.com/github/biothings/biothings_explorer/blob/master/jupyter%20notebooks/EXPLAIN_demo.ipynb).**"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Step 0: Load BioThings Explorer modules"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "First, install the `biothings_explorer` and `biothings_schema` packages, as described in this [README](https://github.com/biothings/biothings_explorer/blob/master/jupyter%20notebooks/README.md#prerequisite).  This only needs to be done once (but including it here for compability with [colab](https://colab.research.google.com/))."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Requirement already satisfied: biothings_explorer from git+https://github.com/biothings/biothings_explorer#egg=biothings_explorer in /home/asu/env/bte_schema/lib/python3.6/site-packages (0.0.1)\n",
      "Requirement already satisfied: jupyter in /home/asu/env/bte_schema/lib/python3.6/site-packages (from biothings_explorer) (1.0.0)\n",
      "Requirement already satisfied: tornado==4.5.3 in /home/asu/env/bte_schema/lib/python3.6/site-packages (from biothings_explorer) (4.5.3)\n",
      "Requirement already satisfied: jsonpath-rw>=1.4.0 in /home/asu/env/bte_schema/lib/python3.6/site-packages (from biothings_explorer) (1.4.0)\n",
      "Requirement already satisfied: notebook==5.7.5 in /home/asu/env/bte_schema/lib/python3.6/site-packages (from biothings_explorer) (5.7.5)\n",
      "Requirement already satisfied: requests>=2.21.0 in /home/asu/env/bte_schema/lib/python3.6/site-packages (from biothings_explorer) (2.22.0)\n",
      "Requirement already satisfied: graphviz>=0.11.1 in /home/asu/env/bte_schema/lib/python3.6/site-packages (from biothings_explorer) (0.13)\n",
      "Requirement already satisfied: jsonschema>=3.0.1 in /home/asu/env/bte_schema/lib/python3.6/site-packages (from biothings_explorer) (3.0.2)\n",
      "Requirement already satisfied: pandas in /home/asu/env/bte_schema/lib/python3.6/site-packages (from biothings_explorer) (0.25.1)\n",
      "Requirement already satisfied: aiohttp in /home/asu/env/bte_schema/lib/python3.6/site-packages (from biothings_explorer) (3.6.0)\n",
      "Requirement already satisfied: networkx==2.3 in /home/asu/env/bte_schema/lib/python3.6/site-packages (from biothings_explorer) (2.3)\n",
      "Requirement already satisfied: jupyter-console in /home/asu/env/bte_schema/lib/python3.6/site-packages (from jupyter->biothings_explorer) (6.0.0)\n",
      "Requirement already satisfied: nbconvert in /home/asu/env/bte_schema/lib/python3.6/site-packages (from jupyter->biothings_explorer) (5.6.0)\n",
      "Requirement already satisfied: ipykernel in /home/asu/env/bte_schema/lib/python3.6/site-packages (from jupyter->biothings_explorer) (5.1.2)\n",
      "Requirement already satisfied: ipywidgets in /home/asu/env/bte_schema/lib/python3.6/site-packages (from jupyter->biothings_explorer) (7.5.1)\n",
      "Requirement already satisfied: qtconsole in /home/asu/env/bte_schema/lib/python3.6/site-packages (from jupyter->biothings_explorer) (4.5.5)\n",
      "Requirement already satisfied: decorator in /home/asu/env/bte_schema/lib/python3.6/site-packages (from jsonpath-rw>=1.4.0->biothings_explorer) (4.4.0)\n",
      "Requirement already satisfied: six in /home/asu/env/bte_schema/lib/python3.6/site-packages (from jsonpath-rw>=1.4.0->biothings_explorer) (1.12.0)\n",
      "Requirement already satisfied: ply in /home/asu/env/bte_schema/lib/python3.6/site-packages (from jsonpath-rw>=1.4.0->biothings_explorer) (3.11)\n",
      "Requirement already satisfied: Send2Trash in /home/asu/env/bte_schema/lib/python3.6/site-packages (from notebook==5.7.5->biothings_explorer) (1.5.0)\n",
      "Requirement already satisfied: pyzmq>=17 in /home/asu/env/bte_schema/lib/python3.6/site-packages (from notebook==5.7.5->biothings_explorer) (18.1.0)\n",
      "Requirement already satisfied: jupyter-core>=4.4.0 in /home/asu/env/bte_schema/lib/python3.6/site-packages (from notebook==5.7.5->biothings_explorer) (4.5.0)\n",
      "Requirement already satisfied: nbformat in /home/asu/env/bte_schema/lib/python3.6/site-packages (from notebook==5.7.5->biothings_explorer) (4.4.0)\n",
      "Requirement already satisfied: terminado>=0.8.1 in /home/asu/env/bte_schema/lib/python3.6/site-packages (from notebook==5.7.5->biothings_explorer) (0.8.2)\n",
      "Requirement already satisfied: prometheus-client in /home/asu/env/bte_schema/lib/python3.6/site-packages (from notebook==5.7.5->biothings_explorer) (0.7.1)\n",
      "Requirement already satisfied: jinja2 in /home/asu/env/bte_schema/lib/python3.6/site-packages (from notebook==5.7.5->biothings_explorer) (2.10.1)\n",
      "Requirement already satisfied: jupyter-client>=5.2.0 in /home/asu/env/bte_schema/lib/python3.6/site-packages (from notebook==5.7.5->biothings_explorer) (5.3.1)\n",
      "Requirement already satisfied: traitlets>=4.2.1 in /home/asu/env/bte_schema/lib/python3.6/site-packages (from notebook==5.7.5->biothings_explorer) (4.3.2)\n",
      "Requirement already satisfied: ipython-genutils in /home/asu/env/bte_schema/lib/python3.6/site-packages (from notebook==5.7.5->biothings_explorer) (0.2.0)\n",
      "Requirement already satisfied: idna<2.9,>=2.5 in /home/asu/env/bte_schema/lib/python3.6/site-packages (from requests>=2.21.0->biothings_explorer) (2.8)\n",
      "Requirement already satisfied: chardet<3.1.0,>=3.0.2 in /home/asu/env/bte_schema/lib/python3.6/site-packages (from requests>=2.21.0->biothings_explorer) (3.0.4)\n",
      "Requirement already satisfied: urllib3!=1.25.0,!=1.25.1,<1.26,>=1.21.1 in /home/asu/env/bte_schema/lib/python3.6/site-packages (from requests>=2.21.0->biothings_explorer) (1.25.3)\n",
      "Requirement already satisfied: certifi>=2017.4.17 in /home/asu/env/bte_schema/lib/python3.6/site-packages (from requests>=2.21.0->biothings_explorer) (2019.9.11)\n",
      "Requirement already satisfied: pyrsistent>=0.14.0 in /home/asu/env/bte_schema/lib/python3.6/site-packages (from jsonschema>=3.0.1->biothings_explorer) (0.15.4)\n",
      "Requirement already satisfied: setuptools in /home/asu/env/bte_schema/lib/python3.6/site-packages (from jsonschema>=3.0.1->biothings_explorer) (41.2.0)\n",
      "Requirement already satisfied: attrs>=17.4.0 in /home/asu/env/bte_schema/lib/python3.6/site-packages (from jsonschema>=3.0.1->biothings_explorer) (19.1.0)\n",
      "Requirement already satisfied: numpy>=1.13.3 in /home/asu/env/bte_schema/lib/python3.6/site-packages (from pandas->biothings_explorer) (1.17.2)\n",
      "Requirement already satisfied: python-dateutil>=2.6.1 in /home/asu/env/bte_schema/lib/python3.6/site-packages (from pandas->biothings_explorer) (2.8.0)\n",
      "Requirement already satisfied: pytz>=2017.2 in /home/asu/env/bte_schema/lib/python3.6/site-packages (from pandas->biothings_explorer) (2019.2)\n",
      "Requirement already satisfied: multidict<5.0,>=4.0 in /home/asu/env/bte_schema/lib/python3.6/site-packages (from aiohttp->biothings_explorer) (4.5.2)\n",
      "Requirement already satisfied: async-timeout<4.0,>=3.0 in /home/asu/env/bte_schema/lib/python3.6/site-packages (from aiohttp->biothings_explorer) (3.0.1)\n",
      "Requirement already satisfied: idna-ssl>=1.0; python_version < \"3.7\" in /home/asu/env/bte_schema/lib/python3.6/site-packages (from aiohttp->biothings_explorer) (1.1.0)\n",
      "Requirement already satisfied: typing-extensions>=3.6.5; python_version < \"3.7\" in /home/asu/env/bte_schema/lib/python3.6/site-packages (from aiohttp->biothings_explorer) (3.7.4)\n",
      "Requirement already satisfied: yarl<2.0,>=1.0 in /home/asu/env/bte_schema/lib/python3.6/site-packages (from aiohttp->biothings_explorer) (1.3.0)\n",
      "Requirement already satisfied: prompt-toolkit<2.1.0,>=2.0.0 in /home/asu/env/bte_schema/lib/python3.6/site-packages (from jupyter-console->jupyter->biothings_explorer) (2.0.9)\n",
      "Requirement already satisfied: pygments in /home/asu/env/bte_schema/lib/python3.6/site-packages (from jupyter-console->jupyter->biothings_explorer) (2.4.2)\n",
      "Requirement already satisfied: ipython in /home/asu/env/bte_schema/lib/python3.6/site-packages (from jupyter-console->jupyter->biothings_explorer) (7.8.0)\n",
      "Requirement already satisfied: pandocfilters>=1.4.1 in /home/asu/env/bte_schema/lib/python3.6/site-packages (from nbconvert->jupyter->biothings_explorer) (1.4.2)\n",
      "Requirement already satisfied: entrypoints>=0.2.2 in /home/asu/env/bte_schema/lib/python3.6/site-packages (from nbconvert->jupyter->biothings_explorer) (0.3)\n",
      "Requirement already satisfied: testpath in /home/asu/env/bte_schema/lib/python3.6/site-packages (from nbconvert->jupyter->biothings_explorer) (0.4.2)\n",
      "Requirement already satisfied: mistune<2,>=0.8.1 in /home/asu/env/bte_schema/lib/python3.6/site-packages (from nbconvert->jupyter->biothings_explorer) (0.8.4)\n",
      "Requirement already satisfied: defusedxml in /home/asu/env/bte_schema/lib/python3.6/site-packages (from nbconvert->jupyter->biothings_explorer) (0.6.0)\n",
      "Requirement already satisfied: bleach in /home/asu/env/bte_schema/lib/python3.6/site-packages (from nbconvert->jupyter->biothings_explorer) (3.1.0)\n",
      "Requirement already satisfied: widgetsnbextension~=3.5.0 in /home/asu/env/bte_schema/lib/python3.6/site-packages (from ipywidgets->jupyter->biothings_explorer) (3.5.1)\n",
      "Requirement already satisfied: ptyprocess; os_name != \"nt\" in /home/asu/env/bte_schema/lib/python3.6/site-packages (from terminado>=0.8.1->notebook==5.7.5->biothings_explorer) (0.6.0)\n",
      "Requirement already satisfied: MarkupSafe>=0.23 in /home/asu/env/bte_schema/lib/python3.6/site-packages (from jinja2->notebook==5.7.5->biothings_explorer) (1.1.1)\n",
      "Requirement already satisfied: wcwidth in /home/asu/env/bte_schema/lib/python3.6/site-packages (from prompt-toolkit<2.1.0,>=2.0.0->jupyter-console->jupyter->biothings_explorer) (0.1.7)\n",
      "Requirement already satisfied: pexpect; sys_platform != \"win32\" in /home/asu/env/bte_schema/lib/python3.6/site-packages (from ipython->jupyter-console->jupyter->biothings_explorer) (4.7.0)\n",
      "Requirement already satisfied: backcall in /home/asu/env/bte_schema/lib/python3.6/site-packages (from ipython->jupyter-console->jupyter->biothings_explorer) (0.1.0)\n",
      "Requirement already satisfied: pickleshare in /home/asu/env/bte_schema/lib/python3.6/site-packages (from ipython->jupyter-console->jupyter->biothings_explorer) (0.7.5)\n",
      "Requirement already satisfied: jedi>=0.10 in /home/asu/env/bte_schema/lib/python3.6/site-packages (from ipython->jupyter-console->jupyter->biothings_explorer) (0.15.1)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Requirement already satisfied: webencodings in /home/asu/env/bte_schema/lib/python3.6/site-packages (from bleach->nbconvert->jupyter->biothings_explorer) (0.5.1)\n",
      "Requirement already satisfied: parso>=0.5.0 in /home/asu/env/bte_schema/lib/python3.6/site-packages (from jedi>=0.10->ipython->jupyter-console->jupyter->biothings_explorer) (0.5.1)\n",
      "Requirement already satisfied: biothings_schema from git+https://github.com/biothings/biothings_schema.py#egg=biothings_schema in /home/asu/env/bte_schema/lib/python3.6/site-packages (0.0.1)\n",
      "Requirement already satisfied: networkx>=2.3 in /home/asu/env/bte_schema/lib/python3.6/site-packages (from biothings_schema) (2.3)\n",
      "Requirement already satisfied: strict-rfc3339 in /home/asu/env/bte_schema/lib/python3.6/site-packages (from biothings_schema) (0.7)\n",
      "Requirement already satisfied: requests>=2.21.0 in /home/asu/env/bte_schema/lib/python3.6/site-packages (from biothings_schema) (2.22.0)\n",
      "Requirement already satisfied: jsonschema>=3.0.1 in /home/asu/env/bte_schema/lib/python3.6/site-packages (from biothings_schema) (3.0.2)\n",
      "Requirement already satisfied: PyYAML>=5.1 in /home/asu/env/bte_schema/lib/python3.6/site-packages (from biothings_schema) (5.1.2)\n",
      "Requirement already satisfied: decorator>=4.3.0 in /home/asu/env/bte_schema/lib/python3.6/site-packages (from networkx>=2.3->biothings_schema) (4.4.0)\n",
      "Requirement already satisfied: idna<2.9,>=2.5 in /home/asu/env/bte_schema/lib/python3.6/site-packages (from requests>=2.21.0->biothings_schema) (2.8)\n",
      "Requirement already satisfied: certifi>=2017.4.17 in /home/asu/env/bte_schema/lib/python3.6/site-packages (from requests>=2.21.0->biothings_schema) (2019.9.11)\n",
      "Requirement already satisfied: urllib3!=1.25.0,!=1.25.1,<1.26,>=1.21.1 in /home/asu/env/bte_schema/lib/python3.6/site-packages (from requests>=2.21.0->biothings_schema) (1.25.3)\n",
      "Requirement already satisfied: chardet<3.1.0,>=3.0.2 in /home/asu/env/bte_schema/lib/python3.6/site-packages (from requests>=2.21.0->biothings_schema) (3.0.4)\n",
      "Requirement already satisfied: setuptools in /home/asu/env/bte_schema/lib/python3.6/site-packages (from jsonschema>=3.0.1->biothings_schema) (41.2.0)\n",
      "Requirement already satisfied: attrs>=17.4.0 in /home/asu/env/bte_schema/lib/python3.6/site-packages (from jsonschema>=3.0.1->biothings_schema) (19.1.0)\n",
      "Requirement already satisfied: six>=1.11.0 in /home/asu/env/bte_schema/lib/python3.6/site-packages (from jsonschema>=3.0.1->biothings_schema) (1.12.0)\n",
      "Requirement already satisfied: pyrsistent>=0.14.0 in /home/asu/env/bte_schema/lib/python3.6/site-packages (from jsonschema>=3.0.1->biothings_schema) (0.15.4)\n"
     ]
    }
   ],
   "source": [
    "!pip install git+https://github.com/biothings/biothings_explorer#egg=biothings_explorer\n",
    "!pip install git+https://github.com/biothings/biothings_schema.py#egg=biothings_schema"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Next, import the relevant modules:\n",
    "\n",
    "* **Hint**: Find corresponding bio-entity representation used in BioThings Explorer based on user input (could be any database IDs, symbols, names)\n",
    "* **FindConnection**: Find intermediate bio-entities which connects user specified input and output"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# import modules from biothings_explorer\n",
    "from biothings_explorer.hint import Hint\n",
    "from biothings_explorer.user_query_dispatcher import FindConnection"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Step 1: Find representation of \"chronic myelogenous leukemia\" and \"imatinib\" in BTE\n",
    "\n",
    "In this step, BioThings Explorer translates our query strings \"chronic myelogenous leukemia\" and \"imatinib\" into BioThings objects, which contain mappings to many common identifiers.  Generally, the top result returned by the `Hint` module will be the correct item, but you should confirm that using the identifiers shown.\n",
    "\n",
    "Search terms can correspond to any child of [BiologicalEntity](https://biolink.github.io/biolink-model/docs/BiologicalEntity.html) from the [Biolink Model](https://biolink.github.io/biolink-model/docs/), including `DiseaseOrPhenotypicFeature` (e.g., \"lupus\"), `ChemicalSubstance` (e.g., \"acetaminophen\"), `Gene` (e.g., \"CDK2\"), `BiologicalProcess` (e.g., \"T cell differentiation\"), and `Pathway` (e.g., \"Citric acid cycle\")."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'mondo': 'MONDO:0011996',\n",
       " 'doid': 'DOID:8552',\n",
       " 'umls': 'C1292772',\n",
       " 'mesh': 'D015464',\n",
       " 'name': 'chronic myelogenous leukemia',\n",
       " 'display': 'mondo(MONDO:0011996) doid(DOID:8552) umls(C1292772) mesh(D015464) name(chronic myelogenous leukemia) ',\n",
       " 'type': 'DiseaseOrPhenotypicFeature',\n",
       " 'primary': {'identifier': 'mondo',\n",
       "  'cls': 'DiseaseOrPhenotypicFeature',\n",
       "  'value': 'MONDO:0011996'}}"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ht = Hint()\n",
    "# find all potential representations of CML\n",
    "cml_hint = ht.query(\"chronic myelogenous leukemia\")\n",
    "# select the correct representation of CML\n",
    "cml = cml_hint['DiseaseOrPhenotypicFeature'][0]\n",
    "cml"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'chembl': 'CHEMBL941',\n",
       " 'drugbank': 'DB00619',\n",
       " 'name': 'IMATINIB',\n",
       " 'pubchem': 5291,\n",
       " 'umls': 'C0935989',\n",
       " 'mesh': 'D000068877',\n",
       " 'display': 'chembl(CHEMBL941) drugbank(DB00619) name(IMATINIB) pubchem(5291) umls(C0935989) mesh(D000068877) ',\n",
       " 'type': 'ChemicalSubstance',\n",
       " 'primary': {'identifier': 'chembl',\n",
       "  'cls': 'ChemicalSubstance',\n",
       "  'value': 'CHEMBL941'}}"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# find all potential representations of imatinib\n",
    "imatinib_hint = ht.query(\"imatinib\")\n",
    "# select the correct representation of imatinib\n",
    "imatinib = imatinib_hint['ChemicalSubstance'][0]\n",
    "imatinib"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Step 2: Find intermediate nodes connecting imatinib and chronic myelogenous leukemia\n",
    "\n",
    "In this section, we find all paths in the knowledge graph that connect imatinib and chronic myelogenous leukemia.  To do that, we will use `FindConnection`.  This class is a convenient wrapper around two advanced functions for **query path planning** and **query path execution**. More advanced features for both query path planning and query path execution are in development and will be documented in the coming months. \n",
    "\n",
    "The parameters for `FindConnection` are described below:\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Help on function __init__ in module biothings_explorer.user_query_dispatcher:\n",
      "\n",
      "__init__(self, input_obj, output_obj, intermediate_nodes, registry=None)\n",
      "    Find relationships in the Knowledge Graph between an Input Object and an Output Object.\n",
      "    \n",
      "    params\n",
      "    ------\n",
      "    input_obj (required): must be an object returned from Hint corresponding to a specific biomedical entity.\n",
      "                          Examples: \n",
      "            Hint().query(\"Fanconi anemia\")['DiseaseOrPhenotypicFeature'][0]\n",
      "            Hint().query(\"acetaminophen\")['ChemicalSubstance'][0]\n",
      "    \n",
      "    output_obj (required): must EITHER be an object returned from Hint corresponding to a specific biomedical\n",
      "                           entity, OR be a string or list of strings corresponding to Biolink Entity classes.\n",
      "                           Examples:\n",
      "            Hint().query(\"acetaminophen\")['ChemicalSubstance'][0]\n",
      "            'Gene'\n",
      "            ['Gene','ChemicalSubstance']\n",
      "    \n",
      "    intermediate_nodes (required): the semantic type(s) of the intermediate node(s).  Examples:\n",
      "            None                         : no intermediate node, find direct connections only\n",
      "            []                           : no intermediate node, find direct connections only\n",
      "            ['BiologicalEntity']         : one intermediate node of any semantic type\n",
      "            ['Gene']                     : one intermediate node that must be a Gene\n",
      "            [('Gene','Pathway')]         : one intermediate node that must be a Gene or a Pathway\n",
      "            ['Gene','Pathway']           : two intermediate nodes, first must be a Gene, second must be a Pathway.\n",
      "            ['Gene',('Pathway','Gene')]  : two intermediate nodes, first must be a Gene, second must be a Pathway or Gene.\n",
      "                                              **NOTE**: queries with more than one intermediate node are currently not supported\n",
      "\n"
     ]
    }
   ],
   "source": [
    "help(FindConnection.__init__)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Here, we formulate a `FindConnection` query with \"CML\" as the `input_ojb`, \"imatinib\" as the `output_obj`.  We further specify with the `intermediate_nodes` parameter that we are looking for paths joining chronic myelogenous leukemia and imatinib with *one* intermediate node that is a Gene.  (The ability to search for longer reasoning paths that include additional intermediate nodes will be added shortly.)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "fc = FindConnection(input_obj=cml, output_obj=imatinib, intermediate_nodes='Gene')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We next execute the `connect` method, which performs the **query path planning** and **query path execution** process.  In short, BioThings Explorer is deconstructing the query into individual API calls, executing those API calls, then assembling the results.\n",
    "\n",
    "A verbose log of this process is displayed below:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "==========\n",
      "========== QUERY PARAMETER SUMMARY ==========\n",
      "==========\n",
      "\n",
      "BTE will find paths that join 'chronic myelogenous leukemia' and 'IMATINIB'. Paths will have 1 intermediate node.\n",
      "\n",
      "Intermediate node #1 will have these type constraints: Gene\n",
      "\n",
      "\n",
      "==========\n",
      "========== QUERY #1 -- fetch all Gene entities linked to 'chronic myelogenous leukemia' ==========\n",
      "==========\n",
      "\n",
      "==== Step #1: Query path planning ====\n",
      "\n",
      "Because chronic myelogenous leukemia is of type 'DiseaseOrPhenotypicFeature', BTE will query our meta-KG for APIs that can take 'DiseaseOrPhenotypicFeature' as input\n",
      "\n",
      "BTE found 3 apis:\n",
      "\n",
      "API 1. mydisease.info(1 API call)\n",
      "API 2. semmeddisease(1 API call)\n",
      "API 3. biolink_disease2gene(1 API call)\n",
      "\n",
      "\n",
      "==== Step #2: Query path execution ====\n",
      "NOTE: API requests are dispatched in parallel, so the list of APIs below is ordered by query time.\n",
      "\n",
      "API 1.1: http://mydisease.info/v1/query (POST \"q=C1292772,C0023473&scopes=mondo.xrefs.umls,disgenet.xrefs.umls&fields=disgenet.genes_related_to_disease&species=human&size=100\")\n",
      "API 2.1: http://pending.biothings.io/semmed/query (POST \"q=C1292772,C0023473&scopes=umls&fields=AFFECTS.protein,AFFECTS_reverse.gene,AFFECTS.gene,CAUSES_reverse.gene,ASSOCIATED_WITH.gene,AFFECTS_reverse.protein&species=human&size=100\")\n",
      "API 3.1: https://api.monarchinitiative.org/api/bioentity/disease/MONDO:0011996/genes?rows=100\n",
      "\n",
      "\n",
      "==== Step #3: Output normalization ====\n",
      "\n",
      "API 3.1 biolink_disease2gene: 4 hits\n",
      "API 1.1 mydisease.info: 52 hits\n",
      "API 2.1 semmeddisease: 160 hits\n",
      "\n",
      "After id-to-object translation, BTE retrieved 143 unique objects.\n",
      "\n",
      "\n",
      "==========\n",
      "========== QUERY #2 -- fetch all Gene entities linked to 'IMATINIB' ==========\n",
      "==========\n",
      "\n",
      "==== Step #1: Query path planning ====\n",
      "\n",
      "Because IMATINIB is of type 'ChemicalSubstance', BTE will query our meta-KG for APIs that can take 'ChemicalSubstance' as input\n",
      "\n",
      "BTE found 3 apis:\n",
      "\n",
      "API 1. mychem.info(2 API calls)\n",
      "API 2. semmedgene(6 API calls)\n",
      "API 3. dgidb_chemical2gene(1 API call)\n",
      "\n",
      "\n",
      "==== Step #2: Query path execution ====\n",
      "NOTE: API requests are dispatched in parallel, so the list of APIs below is ordered by query time.\n",
      "\n",
      "API 1.2: http://mychem.info/v1/query (POST \"q=DB00619&scopes=drugbank.id&fields=drugbank.targets,drugbank.enzymes&species=human&size=100\")\n",
      "API 1.1: http://mychem.info/v1/query (POST \"q=CHEMBL941&scopes=chembl.molecule_chembl_id&fields=drugcentral.bioactivity&species=human&size=100\")\n",
      "API 2.4: https://pending.biothings.io/semmedgene/query (POST \"q=C0935989&scopes=associated_with_reverse.chemical_substance.umls&fields=umls&species=human&size=100\")\n",
      "API 2.1: https://pending.biothings.io/semmedgene/query (POST \"q=C0935989&scopes=affects.chemical_substance.umls&fields=umls&species=human&size=100\")\n",
      "API 2.2: https://pending.biothings.io/semmedgene/query (POST \"q=C0935989&scopes=affects_reverse.chemical_substance.umls&fields=umls&species=human&size=100\")\n",
      "API 2.3: https://pending.biothings.io/semmedgene/query (POST \"q=C0935989&scopes=associated_with.chemical_substance.umls&fields=umls&species=human&size=100\")\n",
      "API 3.1: http://www.dgidb.org/api/v2/interactions.json?drugs=CHEMBL941\n",
      "API 2.6: https://pending.biothings.io/semmedgene/query (POST \"q=C0935989&scopes=interacts_with_reverse.chemical_substance.umls&fields=umls&species=human&size=100\")\n",
      "API 2.5: https://pending.biothings.io/semmedgene/query (POST \"q=C0935989&scopes=interacts_with.chemical_substance.umls&fields=umls&species=human&size=100\")\n",
      "\n",
      "\n",
      "==== Step #3: Output normalization ====\n",
      "\n",
      "API 3.1 dgidb_chemical2gene: 34 hits\n",
      "API 1.1 mychem.info: 77 hits\n",
      "API 1.2 mychem.info: 18 hits\n",
      "API 2.1 semmedgene: No hits\n",
      "API 2.2 semmedgene: No hits\n",
      "API 2.3 semmedgene: No hits\n",
      "API 2.4 semmedgene: No hits\n",
      "API 2.5 semmedgene: 100 hits\n",
      "API 2.6 semmedgene: 100 hits\n",
      "\n",
      "After id-to-object translation, BTE retrieved 244 unique objects.\n",
      "\n",
      "==========\n",
      "========== Final assembly of results ==========\n",
      "==========\n",
      "\n",
      "\n",
      "BTE found 51 unique intermediate nodes connecting 'chronic myelogenous leukemia' and 'IMATINIB'\n"
     ]
    }
   ],
   "source": [
    "# set verbose=True will display all steps which BTE takes to find the connection\n",
    "fc.connect(verbose=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Step 3: Display and Filter results\n",
    "This section demonstrates post-query filtering done in Python. Later, more advanced filtering functions will be added to the **query path execution** module for interleaved filtering, thereby enabling longer query paths. More details to come...\n",
    "\n",
    "First, all matching paths can be exported to a data frame. Let's examine a sample of those results."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>input</th>\n",
       "      <th>input_type</th>\n",
       "      <th>pred1</th>\n",
       "      <th>pred1_source</th>\n",
       "      <th>pred1_api</th>\n",
       "      <th>pred1_pubmed</th>\n",
       "      <th>node1_id</th>\n",
       "      <th>node1_name</th>\n",
       "      <th>node1_type</th>\n",
       "      <th>pred2</th>\n",
       "      <th>pred2_source</th>\n",
       "      <th>pred2_api</th>\n",
       "      <th>pred2_pubmed</th>\n",
       "      <th>output_id</th>\n",
       "      <th>output_name</th>\n",
       "      <th>output_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>30</td>\n",
       "      <td>chronic myelogenous leukemia</td>\n",
       "      <td>DiseaseOrPhenotypicFeature</td>\n",
       "      <td>causedBy</td>\n",
       "      <td>semmed</td>\n",
       "      <td>semmeddisease</td>\n",
       "      <td></td>\n",
       "      <td>entrez:6776</td>\n",
       "      <td>STAT5A</td>\n",
       "      <td>Gene</td>\n",
       "      <td>molecularlyInteractsWith</td>\n",
       "      <td>semmedgene</td>\n",
       "      <td>semmedgene</td>\n",
       "      <td>None</td>\n",
       "      <td>chembl:CHEMBL941</td>\n",
       "      <td>IMATINIB</td>\n",
       "      <td>ChemicalSubstance</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>117</td>\n",
       "      <td>chronic myelogenous leukemia</td>\n",
       "      <td>DiseaseOrPhenotypicFeature</td>\n",
       "      <td>associatedWith</td>\n",
       "      <td>semmed</td>\n",
       "      <td>semmeddisease</td>\n",
       "      <td></td>\n",
       "      <td>entrez:207</td>\n",
       "      <td>AKT1</td>\n",
       "      <td>Gene</td>\n",
       "      <td>molecularlyInteractsWith</td>\n",
       "      <td>semmedgene</td>\n",
       "      <td>semmedgene</td>\n",
       "      <td>None</td>\n",
       "      <td>chembl:CHEMBL941</td>\n",
       "      <td>IMATINIB</td>\n",
       "      <td>ChemicalSubstance</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>34</td>\n",
       "      <td>chronic myelogenous leukemia</td>\n",
       "      <td>DiseaseOrPhenotypicFeature</td>\n",
       "      <td>associatedWith</td>\n",
       "      <td>semmed</td>\n",
       "      <td>semmeddisease</td>\n",
       "      <td></td>\n",
       "      <td>entrez:4547</td>\n",
       "      <td>MTTP</td>\n",
       "      <td>Gene</td>\n",
       "      <td>molecularlyInteractsWith</td>\n",
       "      <td>semmedgene</td>\n",
       "      <td>semmedgene</td>\n",
       "      <td>None</td>\n",
       "      <td>chembl:CHEMBL941</td>\n",
       "      <td>IMATINIB</td>\n",
       "      <td>ChemicalSubstance</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>56</td>\n",
       "      <td>chronic myelogenous leukemia</td>\n",
       "      <td>DiseaseOrPhenotypicFeature</td>\n",
       "      <td>associatedWith</td>\n",
       "      <td>semmed</td>\n",
       "      <td>semmeddisease</td>\n",
       "      <td></td>\n",
       "      <td>entrez:10534</td>\n",
       "      <td>ZNRD2</td>\n",
       "      <td>Gene</td>\n",
       "      <td>molecularlyInteractsWith</td>\n",
       "      <td>semmedgene</td>\n",
       "      <td>semmedgene</td>\n",
       "      <td>None</td>\n",
       "      <td>chembl:CHEMBL941</td>\n",
       "      <td>IMATINIB</td>\n",
       "      <td>ChemicalSubstance</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>113</td>\n",
       "      <td>chronic myelogenous leukemia</td>\n",
       "      <td>DiseaseOrPhenotypicFeature</td>\n",
       "      <td>causedBy</td>\n",
       "      <td>semmed</td>\n",
       "      <td>semmeddisease</td>\n",
       "      <td></td>\n",
       "      <td>entrez:613</td>\n",
       "      <td>BCR</td>\n",
       "      <td>Gene</td>\n",
       "      <td>target</td>\n",
       "      <td>dgidb</td>\n",
       "      <td>dgidb_chemical2gene</td>\n",
       "      <td>None</td>\n",
       "      <td>chembl:CHEMBL941</td>\n",
       "      <td>IMATINIB</td>\n",
       "      <td>ChemicalSubstance</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>57</td>\n",
       "      <td>chronic myelogenous leukemia</td>\n",
       "      <td>DiseaseOrPhenotypicFeature</td>\n",
       "      <td>causedBy</td>\n",
       "      <td>semmed</td>\n",
       "      <td>semmeddisease</td>\n",
       "      <td></td>\n",
       "      <td>entrez:10534</td>\n",
       "      <td>ZNRD2</td>\n",
       "      <td>Gene</td>\n",
       "      <td>molecularlyInteractsWith</td>\n",
       "      <td>semmedgene</td>\n",
       "      <td>semmedgene</td>\n",
       "      <td>None</td>\n",
       "      <td>chembl:CHEMBL941</td>\n",
       "      <td>IMATINIB</td>\n",
       "      <td>ChemicalSubstance</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>27</td>\n",
       "      <td>chronic myelogenous leukemia</td>\n",
       "      <td>DiseaseOrPhenotypicFeature</td>\n",
       "      <td>causedBy</td>\n",
       "      <td>semmed</td>\n",
       "      <td>semmeddisease</td>\n",
       "      <td></td>\n",
       "      <td>entrez:6776</td>\n",
       "      <td>STAT5A</td>\n",
       "      <td>Gene</td>\n",
       "      <td>molecularlyInteractsWith</td>\n",
       "      <td>semmedgene</td>\n",
       "      <td>semmedgene</td>\n",
       "      <td>None</td>\n",
       "      <td>chembl:CHEMBL941</td>\n",
       "      <td>IMATINIB</td>\n",
       "      <td>ChemicalSubstance</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>101</td>\n",
       "      <td>chronic myelogenous leukemia</td>\n",
       "      <td>DiseaseOrPhenotypicFeature</td>\n",
       "      <td>associatedWith</td>\n",
       "      <td>biolink</td>\n",
       "      <td>biolink_disease2gene</td>\n",
       "      <td>None</td>\n",
       "      <td>entrez:613</td>\n",
       "      <td>BCR</td>\n",
       "      <td>Gene</td>\n",
       "      <td>target</td>\n",
       "      <td>dgidb</td>\n",
       "      <td>dgidb_chemical2gene</td>\n",
       "      <td>None</td>\n",
       "      <td>chembl:CHEMBL941</td>\n",
       "      <td>IMATINIB</td>\n",
       "      <td>ChemicalSubstance</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>chronic myelogenous leukemia</td>\n",
       "      <td>DiseaseOrPhenotypicFeature</td>\n",
       "      <td>associatedWith</td>\n",
       "      <td>biolink</td>\n",
       "      <td>biolink_disease2gene</td>\n",
       "      <td></td>\n",
       "      <td>entrez:25</td>\n",
       "      <td>ABL1</td>\n",
       "      <td>Gene</td>\n",
       "      <td>target</td>\n",
       "      <td>dgidb</td>\n",
       "      <td>dgidb_chemical2gene</td>\n",
       "      <td>None</td>\n",
       "      <td>chembl:CHEMBL941</td>\n",
       "      <td>IMATINIB</td>\n",
       "      <td>ChemicalSubstance</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>13</td>\n",
       "      <td>chronic myelogenous leukemia</td>\n",
       "      <td>DiseaseOrPhenotypicFeature</td>\n",
       "      <td>associatedWith</td>\n",
       "      <td>disgenet</td>\n",
       "      <td>mydisease.info</td>\n",
       "      <td>None</td>\n",
       "      <td>entrez:25</td>\n",
       "      <td>ABL1</td>\n",
       "      <td>Gene</td>\n",
       "      <td>target</td>\n",
       "      <td>drugbank</td>\n",
       "      <td>mychem.info</td>\n",
       "      <td>15799618,15917650,15949566,16153117,16205964</td>\n",
       "      <td>chembl:CHEMBL941</td>\n",
       "      <td>IMATINIB</td>\n",
       "      <td>ChemicalSubstance</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                            input                  input_type           pred1  \\\n",
       "30   chronic myelogenous leukemia  DiseaseOrPhenotypicFeature        causedBy   \n",
       "117  chronic myelogenous leukemia  DiseaseOrPhenotypicFeature  associatedWith   \n",
       "34   chronic myelogenous leukemia  DiseaseOrPhenotypicFeature  associatedWith   \n",
       "56   chronic myelogenous leukemia  DiseaseOrPhenotypicFeature  associatedWith   \n",
       "113  chronic myelogenous leukemia  DiseaseOrPhenotypicFeature        causedBy   \n",
       "57   chronic myelogenous leukemia  DiseaseOrPhenotypicFeature        causedBy   \n",
       "27   chronic myelogenous leukemia  DiseaseOrPhenotypicFeature        causedBy   \n",
       "101  chronic myelogenous leukemia  DiseaseOrPhenotypicFeature  associatedWith   \n",
       "4    chronic myelogenous leukemia  DiseaseOrPhenotypicFeature  associatedWith   \n",
       "13   chronic myelogenous leukemia  DiseaseOrPhenotypicFeature  associatedWith   \n",
       "\n",
       "    pred1_source             pred1_api pred1_pubmed      node1_id node1_name  \\\n",
       "30        semmed         semmeddisease                entrez:6776     STAT5A   \n",
       "117       semmed         semmeddisease                 entrez:207       AKT1   \n",
       "34        semmed         semmeddisease                entrez:4547       MTTP   \n",
       "56        semmed         semmeddisease               entrez:10534      ZNRD2   \n",
       "113       semmed         semmeddisease                 entrez:613        BCR   \n",
       "57        semmed         semmeddisease               entrez:10534      ZNRD2   \n",
       "27        semmed         semmeddisease                entrez:6776     STAT5A   \n",
       "101      biolink  biolink_disease2gene         None    entrez:613        BCR   \n",
       "4        biolink  biolink_disease2gene                  entrez:25       ABL1   \n",
       "13      disgenet        mydisease.info         None     entrez:25       ABL1   \n",
       "\n",
       "    node1_type                     pred2 pred2_source            pred2_api  \\\n",
       "30        Gene  molecularlyInteractsWith   semmedgene           semmedgene   \n",
       "117       Gene  molecularlyInteractsWith   semmedgene           semmedgene   \n",
       "34        Gene  molecularlyInteractsWith   semmedgene           semmedgene   \n",
       "56        Gene  molecularlyInteractsWith   semmedgene           semmedgene   \n",
       "113       Gene                    target        dgidb  dgidb_chemical2gene   \n",
       "57        Gene  molecularlyInteractsWith   semmedgene           semmedgene   \n",
       "27        Gene  molecularlyInteractsWith   semmedgene           semmedgene   \n",
       "101       Gene                    target        dgidb  dgidb_chemical2gene   \n",
       "4         Gene                    target        dgidb  dgidb_chemical2gene   \n",
       "13        Gene                    target     drugbank          mychem.info   \n",
       "\n",
       "                                     pred2_pubmed         output_id  \\\n",
       "30                                           None  chembl:CHEMBL941   \n",
       "117                                          None  chembl:CHEMBL941   \n",
       "34                                           None  chembl:CHEMBL941   \n",
       "56                                           None  chembl:CHEMBL941   \n",
       "113                                          None  chembl:CHEMBL941   \n",
       "57                                           None  chembl:CHEMBL941   \n",
       "27                                           None  chembl:CHEMBL941   \n",
       "101                                          None  chembl:CHEMBL941   \n",
       "4                                            None  chembl:CHEMBL941   \n",
       "13   15799618,15917650,15949566,16153117,16205964  chembl:CHEMBL941   \n",
       "\n",
       "    output_name        output_type  \n",
       "30     IMATINIB  ChemicalSubstance  \n",
       "117    IMATINIB  ChemicalSubstance  \n",
       "34     IMATINIB  ChemicalSubstance  \n",
       "56     IMATINIB  ChemicalSubstance  \n",
       "113    IMATINIB  ChemicalSubstance  \n",
       "57     IMATINIB  ChemicalSubstance  \n",
       "27     IMATINIB  ChemicalSubstance  \n",
       "101    IMATINIB  ChemicalSubstance  \n",
       "4      IMATINIB  ChemicalSubstance  \n",
       "13     IMATINIB  ChemicalSubstance  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = fc.display_table_view()\n",
    "\n",
    "# because UMLS is not currently well-integrated in our ID-to-object translation system, removing UMLS-only entries here\n",
    "patternDel = \"^umls:C\\d+\"\n",
    "filter = df.node1_id.str.contains(patternDel)\n",
    "df = df[~filter]\n",
    "\n",
    "df.sample(10)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "While most results are based on edges from [semmed](https://skr3.nlm.nih.gov/SemMed/), edges from [DGIdb](http://www.dgidb.org/), [biolink](https://monarchinitiative.org/), [disgenet](http://www.disgenet.org/), [mydisease.info](https://mydisease.info) and [drugcentral](http://drugcentral.org/) were also retrieved from their respective APIs.  \n",
    "\n",
    "Next, let's look to see which genes are mentioned the most."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "ABL1      20\n",
       "BCR       15\n",
       "KIT       10\n",
       "ABCB1      8\n",
       "AKT1       4\n",
       "STAT5A     4\n",
       "BIRC5      4\n",
       "JAK2       3\n",
       "DCC        2\n",
       "TMED7      2\n",
       "Name: node1_name, dtype: int64"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.node1_name.value_counts().head(10)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Not surprisingly, the top two genes that BioThings Explorer found that join imatinib to CML are *ABL1* and *BCR*, the two genes that are fused in the \"Philadelphia chromosome\", the genetic abnormality that underlies CML, *and* the validate target of imatinib.\n",
    "\n",
    "Let's examine some of the PubMed articles linking **CML to *ABL1*** and ***ABL1* to imatinib**."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "There are 63 articles supporting the edge between CML and ABL1. Sampling of 10 of those:\n",
      "http://pubmed.gov/12623848\n",
      "http://pubmed.gov/20537386\n",
      "http://pubmed.gov/23676790\n",
      "http://pubmed.gov/11964322\n",
      "http://pubmed.gov/22870928\n",
      "http://pubmed.gov/11423618\n",
      "http://pubmed.gov/20010464\n",
      "http://pubmed.gov/24109527\n",
      "http://pubmed.gov/21562040\n",
      "http://pubmed.gov/20557306\n"
     ]
    }
   ],
   "source": [
    "articles = fc.display_edge_info('chronic myelogenous leukemia', 'ABL1')[0]['info']['bts:pubmed']\n",
    "print(\"There are \"+str(len(articles))+\" articles supporting the edge between CML and ABL1. Sampling of 10 of those:\")\n",
    "x = [print(\"http://pubmed.gov/\"+str(x)) for x in articles[0:10] ]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "There are 16 articles supporting the edge between ABL1 and imatinib. Sampling of 10 of those:\n",
      "http://pubmed.gov/11287973\n",
      "http://pubmed.gov/20537386\n",
      "http://pubmed.gov/26758680\n",
      "http://pubmed.gov/20525993\n",
      "http://pubmed.gov/23502220\n",
      "http://pubmed.gov/19075254\n",
      "http://pubmed.gov/15930265\n",
      "http://pubmed.gov/17496201\n",
      "http://pubmed.gov/12637609\n",
      "http://pubmed.gov/11287972\n"
     ]
    }
   ],
   "source": [
    "articles = fc.display_edge_info('ABL1', 'IMATINIB')[4]['info']['bts:publication']\n",
    "print(\"There are \"+str(len(articles))+\" articles supporting the edge between ABL1 and imatinib. Sampling of 10 of those:\")\n",
    "x = [print(\"http://pubmed.gov/\"+str(x)) for x in articles[0:10] ]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Comparing results between CML and GIST\n",
    "\n",
    "Let's perform another BioThings Explorer query, this time looking to EXPLAIN the relationship between imatinib and gastrointestinal stromal tumors (GIST), another disease treated by imatinib."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'mondo': 'MONDO:0011719',\n",
       " 'doid': 'DOID:9253',\n",
       " 'umls': 'C3179349',\n",
       " 'mesh': 'D046152',\n",
       " 'name': 'gastrointestinal stromal tumor',\n",
       " 'display': 'mondo(MONDO:0011719) doid(DOID:9253) umls(C3179349) mesh(D046152) name(gastrointestinal stromal tumor) ',\n",
       " 'type': 'DiseaseOrPhenotypicFeature',\n",
       " 'primary': {'identifier': 'mondo',\n",
       "  'cls': 'DiseaseOrPhenotypicFeature',\n",
       "  'value': 'MONDO:0011719'}}"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ht = Hint()\n",
    "# find all potential representations of CML\n",
    "gist_hint = ht.query(\"gastrointestinal stromal tumor\")\n",
    "# select the correct representation of CML\n",
    "gist = gist_hint['DiseaseOrPhenotypicFeature'][0]\n",
    "gist"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "fc = FindConnection(input_obj=gist, output_obj=imatinib, intermediate_nodes='Gene')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "After id-to-object translation, BTE retrieved 30 unique objects.\n",
      "\n",
      "After id-to-object translation, BTE retrieved 244 unique objects.\n"
     ]
    }
   ],
   "source": [
    "fc.connect(verbose=False) # skipping the verbose log here"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>input</th>\n",
       "      <th>input_type</th>\n",
       "      <th>pred1</th>\n",
       "      <th>pred1_source</th>\n",
       "      <th>pred1_api</th>\n",
       "      <th>pred1_pubmed</th>\n",
       "      <th>node1_id</th>\n",
       "      <th>node1_name</th>\n",
       "      <th>node1_type</th>\n",
       "      <th>pred2</th>\n",
       "      <th>pred2_source</th>\n",
       "      <th>pred2_api</th>\n",
       "      <th>pred2_pubmed</th>\n",
       "      <th>output_id</th>\n",
       "      <th>output_name</th>\n",
       "      <th>output_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>35</td>\n",
       "      <td>gastrointestinal stromal tumor</td>\n",
       "      <td>DiseaseOrPhenotypicFeature</td>\n",
       "      <td>associatedWith</td>\n",
       "      <td>semmed</td>\n",
       "      <td>semmeddisease</td>\n",
       "      <td></td>\n",
       "      <td>entrez:3815</td>\n",
       "      <td>KIT</td>\n",
       "      <td>Gene</td>\n",
       "      <td>molecularlyInteractsWith</td>\n",
       "      <td>semmedgene</td>\n",
       "      <td>semmedgene</td>\n",
       "      <td>None</td>\n",
       "      <td>chembl:CHEMBL941</td>\n",
       "      <td>IMATINIB</td>\n",
       "      <td>ChemicalSubstance</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>39</td>\n",
       "      <td>gastrointestinal stromal tumor</td>\n",
       "      <td>DiseaseOrPhenotypicFeature</td>\n",
       "      <td>causedBy</td>\n",
       "      <td>semmed</td>\n",
       "      <td>semmeddisease</td>\n",
       "      <td></td>\n",
       "      <td>entrez:3815</td>\n",
       "      <td>KIT</td>\n",
       "      <td>Gene</td>\n",
       "      <td>target</td>\n",
       "      <td>dgidb</td>\n",
       "      <td>dgidb_chemical2gene</td>\n",
       "      <td>None</td>\n",
       "      <td>chembl:CHEMBL941</td>\n",
       "      <td>IMATINIB</td>\n",
       "      <td>ChemicalSubstance</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>79</td>\n",
       "      <td>gastrointestinal stromal tumor</td>\n",
       "      <td>DiseaseOrPhenotypicFeature</td>\n",
       "      <td>causedBy</td>\n",
       "      <td>semmed</td>\n",
       "      <td>semmeddisease</td>\n",
       "      <td></td>\n",
       "      <td>entrez:5156</td>\n",
       "      <td>PDGFRA</td>\n",
       "      <td>Gene</td>\n",
       "      <td>target</td>\n",
       "      <td>drugcentral</td>\n",
       "      <td>mychem.info</td>\n",
       "      <td>None</td>\n",
       "      <td>chembl:CHEMBL941</td>\n",
       "      <td>IMATINIB</td>\n",
       "      <td>ChemicalSubstance</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>27</td>\n",
       "      <td>gastrointestinal stromal tumor</td>\n",
       "      <td>DiseaseOrPhenotypicFeature</td>\n",
       "      <td>associatedWith</td>\n",
       "      <td>disgenet</td>\n",
       "      <td>mydisease.info</td>\n",
       "      <td>None</td>\n",
       "      <td>entrez:3815</td>\n",
       "      <td>KIT</td>\n",
       "      <td>Gene</td>\n",
       "      <td>target</td>\n",
       "      <td>drugcentral</td>\n",
       "      <td>mychem.info</td>\n",
       "      <td>None</td>\n",
       "      <td>chembl:CHEMBL941</td>\n",
       "      <td>IMATINIB</td>\n",
       "      <td>ChemicalSubstance</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>42</td>\n",
       "      <td>gastrointestinal stromal tumor</td>\n",
       "      <td>DiseaseOrPhenotypicFeature</td>\n",
       "      <td>associatedWith</td>\n",
       "      <td>biolink</td>\n",
       "      <td>biolink_disease2gene</td>\n",
       "      <td></td>\n",
       "      <td>entrez:5156</td>\n",
       "      <td>PDGFRA</td>\n",
       "      <td>Gene</td>\n",
       "      <td>target</td>\n",
       "      <td>drugcentral</td>\n",
       "      <td>mychem.info</td>\n",
       "      <td>None</td>\n",
       "      <td>chembl:CHEMBL941</td>\n",
       "      <td>IMATINIB</td>\n",
       "      <td>ChemicalSubstance</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>45</td>\n",
       "      <td>gastrointestinal stromal tumor</td>\n",
       "      <td>DiseaseOrPhenotypicFeature</td>\n",
       "      <td>associatedWith</td>\n",
       "      <td>biolink</td>\n",
       "      <td>biolink_disease2gene</td>\n",
       "      <td></td>\n",
       "      <td>entrez:5156</td>\n",
       "      <td>PDGFRA</td>\n",
       "      <td>Gene</td>\n",
       "      <td>target</td>\n",
       "      <td>dgidb</td>\n",
       "      <td>dgidb_chemical2gene</td>\n",
       "      <td>None</td>\n",
       "      <td>chembl:CHEMBL941</td>\n",
       "      <td>IMATINIB</td>\n",
       "      <td>ChemicalSubstance</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>57</td>\n",
       "      <td>gastrointestinal stromal tumor</td>\n",
       "      <td>DiseaseOrPhenotypicFeature</td>\n",
       "      <td>associatedWith</td>\n",
       "      <td>disgenet</td>\n",
       "      <td>mydisease.info</td>\n",
       "      <td>None</td>\n",
       "      <td>entrez:5156</td>\n",
       "      <td>PDGFRA</td>\n",
       "      <td>Gene</td>\n",
       "      <td>target</td>\n",
       "      <td>dgidb</td>\n",
       "      <td>dgidb_chemical2gene</td>\n",
       "      <td>None</td>\n",
       "      <td>chembl:CHEMBL941</td>\n",
       "      <td>IMATINIB</td>\n",
       "      <td>ChemicalSubstance</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>85</td>\n",
       "      <td>gastrointestinal stromal tumor</td>\n",
       "      <td>DiseaseOrPhenotypicFeature</td>\n",
       "      <td>causedBy</td>\n",
       "      <td>semmed</td>\n",
       "      <td>semmeddisease</td>\n",
       "      <td></td>\n",
       "      <td>entrez:5154</td>\n",
       "      <td>PDGFA</td>\n",
       "      <td>Gene</td>\n",
       "      <td>molecularlyInteractsWith</td>\n",
       "      <td>semmedgene</td>\n",
       "      <td>semmedgene</td>\n",
       "      <td>None</td>\n",
       "      <td>chembl:CHEMBL941</td>\n",
       "      <td>IMATINIB</td>\n",
       "      <td>ChemicalSubstance</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>78</td>\n",
       "      <td>gastrointestinal stromal tumor</td>\n",
       "      <td>DiseaseOrPhenotypicFeature</td>\n",
       "      <td>causedBy</td>\n",
       "      <td>semmed</td>\n",
       "      <td>semmeddisease</td>\n",
       "      <td></td>\n",
       "      <td>entrez:5156</td>\n",
       "      <td>PDGFRA</td>\n",
       "      <td>Gene</td>\n",
       "      <td>target</td>\n",
       "      <td>drugcentral</td>\n",
       "      <td>mychem.info</td>\n",
       "      <td>None</td>\n",
       "      <td>chembl:CHEMBL941</td>\n",
       "      <td>IMATINIB</td>\n",
       "      <td>ChemicalSubstance</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>gastrointestinal stromal tumor</td>\n",
       "      <td>DiseaseOrPhenotypicFeature</td>\n",
       "      <td>associatedWith</td>\n",
       "      <td>semmed</td>\n",
       "      <td>semmeddisease</td>\n",
       "      <td></td>\n",
       "      <td>entrez:947</td>\n",
       "      <td>CD34</td>\n",
       "      <td>Gene</td>\n",
       "      <td>molecularlyInteractsWith</td>\n",
       "      <td>semmedgene</td>\n",
       "      <td>semmedgene</td>\n",
       "      <td>None</td>\n",
       "      <td>chembl:CHEMBL941</td>\n",
       "      <td>IMATINIB</td>\n",
       "      <td>ChemicalSubstance</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                             input                  input_type  \\\n",
       "35  gastrointestinal stromal tumor  DiseaseOrPhenotypicFeature   \n",
       "39  gastrointestinal stromal tumor  DiseaseOrPhenotypicFeature   \n",
       "79  gastrointestinal stromal tumor  DiseaseOrPhenotypicFeature   \n",
       "27  gastrointestinal stromal tumor  DiseaseOrPhenotypicFeature   \n",
       "42  gastrointestinal stromal tumor  DiseaseOrPhenotypicFeature   \n",
       "45  gastrointestinal stromal tumor  DiseaseOrPhenotypicFeature   \n",
       "57  gastrointestinal stromal tumor  DiseaseOrPhenotypicFeature   \n",
       "85  gastrointestinal stromal tumor  DiseaseOrPhenotypicFeature   \n",
       "78  gastrointestinal stromal tumor  DiseaseOrPhenotypicFeature   \n",
       "2   gastrointestinal stromal tumor  DiseaseOrPhenotypicFeature   \n",
       "\n",
       "             pred1 pred1_source             pred1_api pred1_pubmed  \\\n",
       "35  associatedWith       semmed         semmeddisease                \n",
       "39        causedBy       semmed         semmeddisease                \n",
       "79        causedBy       semmed         semmeddisease                \n",
       "27  associatedWith     disgenet        mydisease.info         None   \n",
       "42  associatedWith      biolink  biolink_disease2gene                \n",
       "45  associatedWith      biolink  biolink_disease2gene                \n",
       "57  associatedWith     disgenet        mydisease.info         None   \n",
       "85        causedBy       semmed         semmeddisease                \n",
       "78        causedBy       semmed         semmeddisease                \n",
       "2   associatedWith       semmed         semmeddisease                \n",
       "\n",
       "       node1_id node1_name node1_type                     pred2 pred2_source  \\\n",
       "35  entrez:3815        KIT       Gene  molecularlyInteractsWith   semmedgene   \n",
       "39  entrez:3815        KIT       Gene                    target        dgidb   \n",
       "79  entrez:5156     PDGFRA       Gene                    target  drugcentral   \n",
       "27  entrez:3815        KIT       Gene                    target  drugcentral   \n",
       "42  entrez:5156     PDGFRA       Gene                    target  drugcentral   \n",
       "45  entrez:5156     PDGFRA       Gene                    target        dgidb   \n",
       "57  entrez:5156     PDGFRA       Gene                    target        dgidb   \n",
       "85  entrez:5154      PDGFA       Gene  molecularlyInteractsWith   semmedgene   \n",
       "78  entrez:5156     PDGFRA       Gene                    target  drugcentral   \n",
       "2    entrez:947       CD34       Gene  molecularlyInteractsWith   semmedgene   \n",
       "\n",
       "              pred2_api pred2_pubmed         output_id output_name  \\\n",
       "35           semmedgene         None  chembl:CHEMBL941    IMATINIB   \n",
       "39  dgidb_chemical2gene         None  chembl:CHEMBL941    IMATINIB   \n",
       "79          mychem.info         None  chembl:CHEMBL941    IMATINIB   \n",
       "27          mychem.info         None  chembl:CHEMBL941    IMATINIB   \n",
       "42          mychem.info         None  chembl:CHEMBL941    IMATINIB   \n",
       "45  dgidb_chemical2gene         None  chembl:CHEMBL941    IMATINIB   \n",
       "57  dgidb_chemical2gene         None  chembl:CHEMBL941    IMATINIB   \n",
       "85           semmedgene         None  chembl:CHEMBL941    IMATINIB   \n",
       "78          mychem.info         None  chembl:CHEMBL941    IMATINIB   \n",
       "2            semmedgene         None  chembl:CHEMBL941    IMATINIB   \n",
       "\n",
       "          output_type  \n",
       "35  ChemicalSubstance  \n",
       "39  ChemicalSubstance  \n",
       "79  ChemicalSubstance  \n",
       "27  ChemicalSubstance  \n",
       "42  ChemicalSubstance  \n",
       "45  ChemicalSubstance  \n",
       "57  ChemicalSubstance  \n",
       "85  ChemicalSubstance  \n",
       "78  ChemicalSubstance  \n",
       "2   ChemicalSubstance  "
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = fc.display_table_view()\n",
    "\n",
    "# because UMLS is not currently well-integrated in our ID-to-object translation system, removing UMLS-only entries here\n",
    "patternDel = \"^umls:C\\d+\"\n",
    "filter = df.node1_id.str.contains(patternDel)\n",
    "df = df[~filter]\n",
    "\n",
    "df.sample(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "PDGFRA    42\n",
       "KIT       25\n",
       "BRAF       9\n",
       "CD34       4\n",
       "BIRC5      2\n",
       "PDGFA      2\n",
       "NCOR1      1\n",
       "SNRPE      1\n",
       "NF1        1\n",
       "VEGFA      1\n",
       "Name: node1_name, dtype: int64"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.node1_name.value_counts().head(10)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Here, the top two genes that BioThings Explorer found that join imatinib to GIST are *PDGFRA* and *KIT*, the most commonly mutated genes found in GIST *and* validated targets of imatinib.\n",
    "\n",
    "While several of the listed genes would be considered positive controls, others on the list could be viewed as **testable hypotheses and discovery opportunities** to be evaluated by domain experts."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Conclusions and caveats\n",
    "\n",
    "This notebook demonstrated the use of BioThings Explorer in EXPLAIN mode to investigate the relationship between imatinib and two diseases that it treats -- chronic myelogenous leukemia (CML) and gastrointestinal stromal tumors (GIST).  In each case, BioThings Explorer autonomously queried a **distributed knowledge graph of biomedical APIs** to find the most common genes, and in each case the relevant targets were retrieved.\n",
    "\n",
    "There are still many areas for improvement (and some areas in which BioThings Explorer is still buggy).  And of course, BioThings Explorer is dependent on the accessibility of the APIs that comprise the distributed knowledge graph.  Nevertheless, we encourage users to try other variants of the EXPLAIN queries demonstrated in this notebook."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
